### PR TITLE
Add node id to configuration and use it instead of endpoint for communicating with brokers

### LINKED ISSUE #1046: A node has an unique id in the cluster
AT:
- The node id can be specified as configuration parameter
- The node id can be set using an environment variable
- Every request which addresses a node internally uses the node id instead of the transport address
- The node id is translated to the transport address using the gossip state

### LINKED ISSUE #1066: Invalid mocking in raft rule
Saw this once when running the five node raft test in a loop

```
[ERROR] shouldRejoinCluster(io.zeebe.raft.RaftFiveNodesTest)  Time elapsed: 24.722 s  <<< ERROR!
java.util.concurrent.ExecutionException: 

Unfinished stubbing detected here:
-> at io.zeebe.raft.util.RaftRule.lambda$reconnectTo$8(RaftRule.java:488)

E.g. thenReturn() may be missing.
Examples of correct stubbing:
    when(mock.isOk()).thenReturn(true);
    when(mock.isOk()).thenThrow(exception);
    doThrow(exception).when(mock).someVoidMethod();
Hints:
 1. missing thenReturn()
 2. you are trying to stub a final method, which is not supported
 3: you are stubbing the behaviour of another mock inside before 'thenReturn' instruction if completed

        at io.zeebe.util.sched.future.CompletableActorFuture.get(CompletableActorFuture.java:152)
        at io.zeebe.util.sched.future.CompletableActorFuture.get(CompletableActorFuture.java:118)
        at io.zeebe.util.sched.FutureUtil.join(FutureUtil.java:28)
        at io.zeebe.util.sched.future.CompletableActorFuture.join(CompletableActorFuture.java:224)
        at io.zeebe.raft.util.RaftRule.reconnectTo(RaftRule.java:498)
        at io.zeebe.raft.util.RaftClusterRule.lambda$registerRaft$0(RaftClusterRule.java:90)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at io.zeebe.raft.util.RaftClusterRule.registerRaft(RaftClusterRule.java:88)
        at io.zeebe.raft.RaftFiveNodesTest.shouldRejoinCluster(RaftFiveNodesTest.java:172)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:236)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:386)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:323)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:143)
Caused by: org.mockito.exceptions.misusing.UnfinishedStubbingException: 
Unfinished stubbing detected here:
-> at io.zeebe.raft.util.RaftRule.lambda$reconnectTo$8(RaftRule.java:488)

E.g. thenReturn() may be missing.
Examples of correct stubbing:
    when(mock.isOk()).thenReturn(true);
    when(mock.isOk()).thenThrow(exception);
    doThrow(exception).when(mock).someVoidMethod();
Hints:
 1. missing thenReturn()
 2. you are trying to stub a final method, which is not supported
 3: you are stubbing the behaviour of another mock inside before 'thenReturn' instruction if completed

        at io.zeebe.raft.util.RaftRule.lambda$reconnectTo$8(RaftRule.java:490)
        at io.zeebe.util.sched.ActorControl.lambda$call$0(ActorControl.java:145)
        at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76)
        at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53)
        at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:187)
        at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154)
        at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135)
        at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112)
```

### PR DESCRIPTION
**Pull request by menski**
The node id should identify a node uniquely in the cluster. It should be between 0 to cluster nodes (exclusive). This allows us to already know on startup all node ids if we are aware of the cluster size.

Transport now contains a component called `EndpointRegistry` which is used to handle the lookup between node id and current endpoint. Endpoint registration and deactivating is happening based on state discovered through gossip. All requests are now send to a node by id instead of a specific endpoint.

There are two special cases where the node id is not known. First on broker bootstrap only the contact points are known without the node id, therefore gossip registers the current tested contact point with a special node id `UNKNOWN_NODE_ID`. After the initial sync request the node ids are known. The same is true for the gateway client which also only has an initial contact point. After the first topology refresh the node ids are known.

Removes as many as possible `host` and `port` fields from the SBE protocol and replaces them with node ids instead. This allows to replace a node by joining a new node with the same id even with different endpoints.

closes #1046 
closes #1066

### CODE REVIEW

File: raft/src/test/java/io/zeebe/raft/util/RaftRule.java [hunk 1]
```diff
@@ -483,60 +403,12 @@ public String toString() {
...
-    int value;
-    try {
-      final Field remoteStreamId = transportMessageClass.getDeclaredField("remoteStreamId");
-      remoteStreamId.setAccessible(true);
-      value = (int) remoteStreamId.get(transportMessage);
-    } catch (Exception e) {
-      value = -1;
-    }
-    return value;
+    clientTransport.deactivateEndpoint(other.getNodeId());
```
Comment (Zelldon) [inline raft/src/test/java/io/zeebe/raft/util/RaftRule.java]: Awesome :)

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]

File: broker-core/src/test/java/io/zeebe/broker/util/BufferingClientOutput.java [hunk 2]
```diff
@@ -83,14 +81,14 @@ public Request getLastRequest() {
 
   public class Request {
     private final int requestId = ID_GEN.incrementAndGet();
-    private final RemoteAddress destination;
+    private final Integer destination;
```
[the remark this answers is not in the document]
Comment (menski) [reply broker-core/src/test/java/io/zeebe/broker/util/BufferingClientOutput.java]: This is a test class and the destination is provided by a `Supplier<Integer>` so could return null. I didn't changed it. Hope that is okay.

**Comment by Zelldon** [comment]
Looks good can we change the other List<Integer> as well to IntArrayList?

**Comment by menski** [comment]
Which list did I missed?

**Comment by menski** [comment]
I squashed and rebased on develop. Waiting for Jenkins build to pass. @Zelldon can I get an approve of the changes to the PR so I can merge it?

**Review by ChrisKujawa — APPROVED** [review]

**Pull request merged**